src/types/record.ts

```diff
@@ -2,26 +2,52 @@ import { Runtype, Static, create, innerValidate } from '../runtype';
 import { hasKey } from '../util';
 import show from '../show';
 
-type RecordStaticType<O extends { [_: string]: Runtype }, RO extends boolean> = RO extends true
+type RecordStaticType<
+  O extends { [_: string]: Runtype },
+  Part extends boolean,
+  RO extends boolean
+> = Part extends true
+  ? RO extends true
+    ? { readonly [K in keyof O]?: Static<O[K]> }
+    : { [K in keyof O]?: Static<O[K]> }
+  : RO extends true
   ? { readonly [K in keyof O]: Static<O[K]> }
   : { [K in keyof O]: Static<O[K]> };
 
-export interface Record<O extends { [_: string]: Runtype }, RO extends boolean>
-  extends Runtype<RecordStaticType<O, RO>> {
+interface InternalRecord<
+  O extends { [_: string]: Runtype },
+  Part extends boolean,
+  RO extends boolean
+> extends Runtype<RecordStaticType<O, Part, RO>> {
   tag: 'record';
   fields: O;
+  isPartial: Part;
   isReadonly: RO;
 
-  asReadonly(): Record<O, true>;
+  asPartial(): InternalRecord<O, true, RO>;
```

Comment:
I threw this in there because it seems like a win for usability.  

The slippery slope includes `asRequired` and `asMutable`.  I'm not sold on those.

Reply:
Fun trick, you could do i.e.

```ts
asPartial(this: { isPartial: false }): InternalRecord<O, true, RO>;
```

and get a type error on when calling this on a `Partial`.